{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final Project  \n",
    "\n",
    "Bioinformatics & Network Medicine  \n",
    "\n",
    "Group 14  \n",
    "\n",
    "Emre Pelzer                  pelzer.2183479@studenti.uniroma1.it  \n",
    "Syed Habibul Bashar          bashar.2102742@studenti.uniroma1.it  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "import itertools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.sparse import csgraph\n",
    "from scipy.stats import hypergeom\n",
    "from sklearn.model_selection import KFold\n",
    "from scipy.sparse.linalg import expm_multiply\n",
    "from sklearn.metrics import precision_score, recall_score, f1_score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Intractome Construction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Syed Habibul Bashar\\AppData\\Local\\Temp\\ipykernel_15192\\3348542099.py:3: DtypeWarning: Columns (1,2,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(file_path, delimiter='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: center;\">\n",
       "      <th></th>\n",
       "      <th>#BioGRID Interaction ID</th>\n",
       "      <th>Entrez Gene Interactor A</th>\n",
       "      <th>Entrez Gene Interactor B</th>\n",
       "      <th>BioGRID ID Interactor A</th>\n",
       "      <th>BioGRID ID Interactor B</th>\n",
       "      <th>Systematic Name Interactor A</th>\n",
       "      <th>Systematic Name Interactor B</th>\n",
       "      <th>Official Symbol Interactor A</th>\n",
       "      <th>Official Symbol Interactor B</th>\n",
       "      <th>Synonyms Interactor A</th>\n",
       "      <th>Synonyms Interactor B</th>\n",
       "      <th>Experimental System</th>\n",
       "      <th>Experimental System Type</th>\n",
       "      <th>Author</th>\n",
       "      <th>Publication Source</th>\n",
       "      <th>Organism ID Interactor A</th>\n",
       "      <th>Organism ID Interactor B</th>\n",
       "      <th>Throughput</th>\n",
       "      <th>Score</th>\n",
       "      <th>Modification</th>\n",
       "      <th>Qualifications</th>\n",
       "      <th>Tags</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>SWISS-PROT Accessions Interactor A</th>\n",
       "      <th>TREMBL Accessions Interactor A</th>\n",
       "      <th>REFSEQ Accessions Interactor A</th>\n",
       "      <th>SWISS-PROT Accessions Interactor B</th>\n",
       "      <th>TREMBL Accessions Interactor B</th>\n",
       "      <th>REFSEQ Accessions Interactor B</th>\n",
       "      <th>Ontology Term IDs</th>\n",
       "      <th>Ontology Term Names</th>\n",
       "      <th>Ontology Term Categories</th>\n",
       "      <th>Ontology Term Qualifier IDs</th>\n",
       "      <th>Ontology Term Qualifier Names</th>\n",
       "      <th>Ontology Term Types</th>\n",
       "      <th>Organism Name Interactor A</th>\n",
       "      <th>Organism Name Interactor B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>103</td>\n",
       "      <td>6416</td>\n",
       "      <td>2318</td>\n",
       "      <td>112315</td>\n",
       "      <td>108607</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MAP2K4</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...</td>\n",
       "      <td>ABP-280|ABP280A|ABPA|ABPL|FLN2|MFM5|MPD4</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Marti A (1997)</td>\n",
       "      <td>PUBMED:9006895</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P45985</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_003001|NP_001268364</td>\n",
       "      <td>Q14315</td>\n",
       "      <td>Q59H94</td>\n",
       "      <td>NP_001120959|NP_001449</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>117</td>\n",
       "      <td>84665</td>\n",
       "      <td>88</td>\n",
       "      <td>124185</td>\n",
       "      <td>106603</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MYPN</td>\n",
       "      <td>ACTN2</td>\n",
       "      <td>CMD1DD|CMH22|MYOP|RCM4</td>\n",
       "      <td>CMD1AA</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Bang ML (2001)</td>\n",
       "      <td>PUBMED:11309420</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q86TC9</td>\n",
       "      <td>A0A087WX60</td>\n",
       "      <td>NP_001243197|NP_001243196|NP_115967</td>\n",
       "      <td>P35609</td>\n",
       "      <td>Q59FD9|F6THM6</td>\n",
       "      <td>NP_001094|NP_001265272|NP_001265273</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>183</td>\n",
       "      <td>90</td>\n",
       "      <td>2339</td>\n",
       "      <td>106605</td>\n",
       "      <td>108625</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>FNTA</td>\n",
       "      <td>ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI</td>\n",
       "      <td>FPTA|PGGT1A|PTAR2</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Wang T (1996)</td>\n",
       "      <td>PUBMED:8599089</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q04771</td>\n",
       "      <td>D3DPA4</td>\n",
       "      <td>NP_001104537|NP_001096</td>\n",
       "      <td>P49354</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_002018</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>278</td>\n",
       "      <td>2624</td>\n",
       "      <td>5371</td>\n",
       "      <td>108894</td>\n",
       "      <td>111384</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "      <td>DCML|IMD21|MONOMAC|NFE1B</td>\n",
       "      <td>MYL|PP8675|RNF71|TRIM19</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Tsuzuki S (2000)</td>\n",
       "      <td>PUBMED:10938104</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P23769</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_001139134|NP_116027|NP_001139133</td>\n",
       "      <td>P29590</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_150250|NP_150253|NP_150252|NP_150247|NP_150...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>418</td>\n",
       "      <td>6118</td>\n",
       "      <td>6774</td>\n",
       "      <td>112038</td>\n",
       "      <td>112651</td>\n",
       "      <td>RP4-547C9.3</td>\n",
       "      <td>-</td>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>REPA2|RP-A p32|RP-A p34|RPA32</td>\n",
       "      <td>ADMIO|APRF|HIES</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Kim J (2000)</td>\n",
       "      <td>PUBMED:10875894</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P15927</td>\n",
       "      <td>B4DUL2</td>\n",
       "      <td>NP_001342057|NP_002937|NP_001284487|NP_0013420...</td>\n",
       "      <td>P40763</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_644805|NP_003141|NP_001356447|NP_001356443|...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #BioGRID Interaction ID Entrez Gene Interactor A Entrez Gene Interactor B  \\\n",
       "0            103                      6416                    2318             \n",
       "1            117                     84665                      88             \n",
       "2            183                        90                    2339             \n",
       "3            278                      2624                    5371             \n",
       "4            418                      6118                    6774             \n",
       "\n",
       "   BioGRID ID Interactor A  BioGRID ID Interactor B  \\\n",
       "0          112315                   108607            \n",
       "1          124185                   106603            \n",
       "2          106605                   108625            \n",
       "3          108894                   111384            \n",
       "4          112038                   112651            \n",
       "\n",
       "  Systematic Name Interactor A Systematic Name Interactor B  \\\n",
       "0                    -                       -                \n",
       "1                    -                       -                \n",
       "2                    -                       -                \n",
       "3                    -                       -                \n",
       "4          RP4-547C9.3                       -                \n",
       "\n",
       "  Official Symbol Interactor A Official Symbol Interactor B  \\\n",
       "0            MAP2K4                         FLNC              \n",
       "1              MYPN                        ACTN2              \n",
       "2             ACVR1                         FNTA              \n",
       "3             GATA2                          PML              \n",
       "4              RPA2                        STAT3              \n",
       "\n",
       "                Synonyms Interactor A                 \\\n",
       "0  JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...   \n",
       "1                             CMD1DD|CMH22|MYOP|RCM4   \n",
       "2            ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI   \n",
       "3                           DCML|IMD21|MONOMAC|NFE1B   \n",
       "4                      REPA2|RP-A p32|RP-A p34|RPA32   \n",
       "\n",
       "            Synonyms Interactor B           Experimental System  \\\n",
       "0  ABP-280|ABP280A|ABPA|ABPL|FLN2|MFM5|MPD4      Two-hybrid       \n",
       "1                                    CMD1AA      Two-hybrid       \n",
       "2                         FPTA|PGGT1A|PTAR2      Two-hybrid       \n",
       "3                   MYL|PP8675|RNF71|TRIM19      Two-hybrid       \n",
       "4                           ADMIO|APRF|HIES      Two-hybrid       \n",
       "\n",
       "  Experimental System Type       Author      Publication Source  \\\n",
       "0         physical            Marti A (1997)    PUBMED:9006895    \n",
       "1         physical            Bang ML (2001)   PUBMED:11309420    \n",
       "2         physical             Wang T (1996)    PUBMED:8599089    \n",
       "3         physical          Tsuzuki S (2000)   PUBMED:10938104    \n",
       "4         physical              Kim J (2000)   PUBMED:10875894    \n",
       "\n",
       "   Organism ID Interactor A  Organism ID Interactor B    Throughput   Score  \\\n",
       "0            9606                      9606            Low Throughput    -    \n",
       "1            9606                      9606            Low Throughput    -    \n",
       "2            9606                      9606            Low Throughput    -    \n",
       "3            9606                      9606            Low Throughput    -    \n",
       "4            9606                      9606            Low Throughput    -    \n",
       "\n",
       "  Modification Qualifications Tags Source Database  \\\n",
       "0       -             -         -       BIOGRID      \n",
       "1       -             -         -       BIOGRID      \n",
       "2       -             -         -       BIOGRID      \n",
       "3       -             -         -       BIOGRID      \n",
       "4       -             -         -       BIOGRID      \n",
       "\n",
       "  SWISS-PROT Accessions Interactor A TREMBL Accessions Interactor A  \\\n",
       "0               P45985                                  -             \n",
       "1               Q86TC9                         A0A087WX60             \n",
       "2               Q04771                             D3DPA4             \n",
       "3               P23769                                  -             \n",
       "4               P15927                             B4DUL2             \n",
       "\n",
       "            REFSEQ Accessions Interactor A            \\\n",
       "0                             NP_003001|NP_001268364   \n",
       "1                NP_001243197|NP_001243196|NP_115967   \n",
       "2                             NP_001104537|NP_001096   \n",
       "3                NP_001139134|NP_116027|NP_001139133   \n",
       "4  NP_001342057|NP_002937|NP_001284487|NP_0013420...   \n",
       "\n",
       "  SWISS-PROT Accessions Interactor B TREMBL Accessions Interactor B  \\\n",
       "0               Q14315                               Q59H94           \n",
       "1               P35609                        Q59FD9|F6THM6           \n",
       "2               P49354                                    -           \n",
       "3               P29590                                    -           \n",
       "4               P40763                                    -           \n",
       "\n",
       "            REFSEQ Accessions Interactor B           Ontology Term IDs  \\\n",
       "0                             NP_001120959|NP_001449          -          \n",
       "1                NP_001094|NP_001265272|NP_001265273          -          \n",
       "2                                          NP_002018          -          \n",
       "3  NP_150250|NP_150253|NP_150252|NP_150247|NP_150...          -          \n",
       "4  NP_644805|NP_003141|NP_001356447|NP_001356443|...          -          \n",
       "\n",
       "  Ontology Term Names Ontology Term Categories Ontology Term Qualifier IDs  \\\n",
       "0           -                     -                          -               \n",
       "1           -                     -                          -               \n",
       "2           -                     -                          -               \n",
       "3           -                     -                          -               \n",
       "4           -                     -                          -               \n",
       "\n",
       "  Ontology Term Qualifier Names Ontology Term Types  \\\n",
       "0                -                        -           \n",
       "1                -                        -           \n",
       "2                -                        -           \n",
       "3                -                        -           \n",
       "4                -                        -           \n",
       "\n",
       "  Organism Name Interactor A Organism Name Interactor B  \n",
       "0        Homo sapiens               Homo sapiens         \n",
       "1        Homo sapiens               Homo sapiens         \n",
       "2        Homo sapiens               Homo sapiens         \n",
       "3        Homo sapiens               Homo sapiens         \n",
       "4        Homo sapiens               Homo sapiens         "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "file_path = r'E:\\University\\BINM\\BIOGRID-ALL-4.4.240.tab3\\BIOGRID-ALL-4.4.240.tab3.txt'\n",
    "\n",
    "df = pd.read_csv(file_path, delimiter='\\t')\n",
    "\n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: center;\">\n",
       "      <th></th>\n",
       "      <th>#BioGRID Interaction ID</th>\n",
       "      <th>Entrez Gene Interactor A</th>\n",
       "      <th>Entrez Gene Interactor B</th>\n",
       "      <th>BioGRID ID Interactor A</th>\n",
       "      <th>BioGRID ID Interactor B</th>\n",
       "      <th>Systematic Name Interactor A</th>\n",
       "      <th>Systematic Name Interactor B</th>\n",
       "      <th>Official Symbol Interactor A</th>\n",
       "      <th>Official Symbol Interactor B</th>\n",
       "      <th>Synonyms Interactor A</th>\n",
       "      <th>Synonyms Interactor B</th>\n",
       "      <th>Experimental System</th>\n",
       "      <th>Experimental System Type</th>\n",
       "      <th>Author</th>\n",
       "      <th>Publication Source</th>\n",
       "      <th>Organism ID Interactor A</th>\n",
       "      <th>Organism ID Interactor B</th>\n",
       "      <th>Throughput</th>\n",
       "      <th>Score</th>\n",
       "      <th>Modification</th>\n",
       "      <th>Qualifications</th>\n",
       "      <th>Tags</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>SWISS-PROT Accessions Interactor A</th>\n",
       "      <th>TREMBL Accessions Interactor A</th>\n",
       "      <th>REFSEQ Accessions Interactor A</th>\n",
       "      <th>SWISS-PROT Accessions Interactor B</th>\n",
       "      <th>TREMBL Accessions Interactor B</th>\n",
       "      <th>REFSEQ Accessions Interactor B</th>\n",
       "      <th>Ontology Term IDs</th>\n",
       "      <th>Ontology Term Names</th>\n",
       "      <th>Ontology Term Categories</th>\n",
       "      <th>Ontology Term Qualifier IDs</th>\n",
       "      <th>Ontology Term Qualifier Names</th>\n",
       "      <th>Ontology Term Types</th>\n",
       "      <th>Organism Name Interactor A</th>\n",
       "      <th>Organism Name Interactor B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>103</td>\n",
       "      <td>6416</td>\n",
       "      <td>2318</td>\n",
       "      <td>112315</td>\n",
       "      <td>108607</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MAP2K4</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...</td>\n",
       "      <td>ABP-280|ABP280A|ABPA|ABPL|FLN2|MFM5|MPD4</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Marti A (1997)</td>\n",
       "      <td>PUBMED:9006895</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P45985</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_003001|NP_001268364</td>\n",
       "      <td>Q14315</td>\n",
       "      <td>Q59H94</td>\n",
       "      <td>NP_001120959|NP_001449</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>117</td>\n",
       "      <td>84665</td>\n",
       "      <td>88</td>\n",
       "      <td>124185</td>\n",
       "      <td>106603</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MYPN</td>\n",
       "      <td>ACTN2</td>\n",
       "      <td>CMD1DD|CMH22|MYOP|RCM4</td>\n",
       "      <td>CMD1AA</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Bang ML (2001)</td>\n",
       "      <td>PUBMED:11309420</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q86TC9</td>\n",
       "      <td>A0A087WX60</td>\n",
       "      <td>NP_001243197|NP_001243196|NP_115967</td>\n",
       "      <td>P35609</td>\n",
       "      <td>Q59FD9|F6THM6</td>\n",
       "      <td>NP_001094|NP_001265272|NP_001265273</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>183</td>\n",
       "      <td>90</td>\n",
       "      <td>2339</td>\n",
       "      <td>106605</td>\n",
       "      <td>108625</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>FNTA</td>\n",
       "      <td>ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI</td>\n",
       "      <td>FPTA|PGGT1A|PTAR2</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Wang T (1996)</td>\n",
       "      <td>PUBMED:8599089</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q04771</td>\n",
       "      <td>D3DPA4</td>\n",
       "      <td>NP_001104537|NP_001096</td>\n",
       "      <td>P49354</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_002018</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>278</td>\n",
       "      <td>2624</td>\n",
       "      <td>5371</td>\n",
       "      <td>108894</td>\n",
       "      <td>111384</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "      <td>DCML|IMD21|MONOMAC|NFE1B</td>\n",
       "      <td>MYL|PP8675|RNF71|TRIM19</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Tsuzuki S (2000)</td>\n",
       "      <td>PUBMED:10938104</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P23769</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_001139134|NP_116027|NP_001139133</td>\n",
       "      <td>P29590</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_150250|NP_150253|NP_150252|NP_150247|NP_150...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>418</td>\n",
       "      <td>6118</td>\n",
       "      <td>6774</td>\n",
       "      <td>112038</td>\n",
       "      <td>112651</td>\n",
       "      <td>RP4-547C9.3</td>\n",
       "      <td>-</td>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>REPA2|RP-A p32|RP-A p34|RPA32</td>\n",
       "      <td>ADMIO|APRF|HIES</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Kim J (2000)</td>\n",
       "      <td>PUBMED:10875894</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P15927</td>\n",
       "      <td>B4DUL2</td>\n",
       "      <td>NP_001342057|NP_002937|NP_001284487|NP_0013420...</td>\n",
       "      <td>P40763</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_644805|NP_003141|NP_001356447|NP_001356443|...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #BioGRID Interaction ID Entrez Gene Interactor A Entrez Gene Interactor B  \\\n",
       "0            103                      6416                    2318             \n",
       "1            117                     84665                      88             \n",
       "2            183                        90                    2339             \n",
       "3            278                      2624                    5371             \n",
       "4            418                      6118                    6774             \n",
       "\n",
       "   BioGRID ID Interactor A  BioGRID ID Interactor B  \\\n",
       "0          112315                   108607            \n",
       "1          124185                   106603            \n",
       "2          106605                   108625            \n",
       "3          108894                   111384            \n",
       "4          112038                   112651            \n",
       "\n",
       "  Systematic Name Interactor A Systematic Name Interactor B  \\\n",
       "0                    -                       -                \n",
       "1                    -                       -                \n",
       "2                    -                       -                \n",
       "3                    -                       -                \n",
       "4          RP4-547C9.3                       -                \n",
       "\n",
       "  Official Symbol Interactor A Official Symbol Interactor B  \\\n",
       "0            MAP2K4                         FLNC              \n",
       "1              MYPN                        ACTN2              \n",
       "2             ACVR1                         FNTA              \n",
       "3             GATA2                          PML              \n",
       "4              RPA2                        STAT3              \n",
       "\n",
       "                Synonyms Interactor A                 \\\n",
       "0  JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...   \n",
       "1                             CMD1DD|CMH22|MYOP|RCM4   \n",
       "2            ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI   \n",
       "3                           DCML|IMD21|MONOMAC|NFE1B   \n",
       "4                      REPA2|RP-A p32|RP-A p34|RPA32   \n",
       "\n",
       "            Synonyms Interactor B           Experimental System  \\\n",
       "0  ABP-280|ABP280A|ABPA|ABPL|FLN2|MFM5|MPD4      Two-hybrid       \n",
       "1                                    CMD1AA      Two-hybrid       \n",
       "2                         FPTA|PGGT1A|PTAR2      Two-hybrid       \n",
       "3                   MYL|PP8675|RNF71|TRIM19      Two-hybrid       \n",
       "4                           ADMIO|APRF|HIES      Two-hybrid       \n",
       "\n",
       "  Experimental System Type       Author      Publication Source  \\\n",
       "0         physical            Marti A (1997)    PUBMED:9006895    \n",
       "1         physical            Bang ML (2001)   PUBMED:11309420    \n",
       "2         physical             Wang T (1996)    PUBMED:8599089    \n",
       "3         physical          Tsuzuki S (2000)   PUBMED:10938104    \n",
       "4         physical              Kim J (2000)   PUBMED:10875894    \n",
       "\n",
       "   Organism ID Interactor A  Organism ID Interactor B    Throughput   Score  \\\n",
       "0            9606                      9606            Low Throughput    -    \n",
       "1            9606                      9606            Low Throughput    -    \n",
       "2            9606                      9606            Low Throughput    -    \n",
       "3            9606                      9606            Low Throughput    -    \n",
       "4            9606                      9606            Low Throughput    -    \n",
       "\n",
       "  Modification Qualifications Tags Source Database  \\\n",
       "0       -             -         -       BIOGRID      \n",
       "1       -             -         -       BIOGRID      \n",
       "2       -             -         -       BIOGRID      \n",
       "3       -             -         -       BIOGRID      \n",
       "4       -             -         -       BIOGRID      \n",
       "\n",
       "  SWISS-PROT Accessions Interactor A TREMBL Accessions Interactor A  \\\n",
       "0               P45985                                  -             \n",
       "1               Q86TC9                         A0A087WX60             \n",
       "2               Q04771                             D3DPA4             \n",
       "3               P23769                                  -             \n",
       "4               P15927                             B4DUL2             \n",
       "\n",
       "            REFSEQ Accessions Interactor A            \\\n",
       "0                             NP_003001|NP_001268364   \n",
       "1                NP_001243197|NP_001243196|NP_115967   \n",
       "2                             NP_001104537|NP_001096   \n",
       "3                NP_001139134|NP_116027|NP_001139133   \n",
       "4  NP_001342057|NP_002937|NP_001284487|NP_0013420...   \n",
       "\n",
       "  SWISS-PROT Accessions Interactor B TREMBL Accessions Interactor B  \\\n",
       "0               Q14315                               Q59H94           \n",
       "1               P35609                        Q59FD9|F6THM6           \n",
       "2               P49354                                    -           \n",
       "3               P29590                                    -           \n",
       "4               P40763                                    -           \n",
       "\n",
       "            REFSEQ Accessions Interactor B           Ontology Term IDs  \\\n",
       "0                             NP_001120959|NP_001449          -          \n",
       "1                NP_001094|NP_001265272|NP_001265273          -          \n",
       "2                                          NP_002018          -          \n",
       "3  NP_150250|NP_150253|NP_150252|NP_150247|NP_150...          -          \n",
       "4  NP_644805|NP_003141|NP_001356447|NP_001356443|...          -          \n",
       "\n",
       "  Ontology Term Names Ontology Term Categories Ontology Term Qualifier IDs  \\\n",
       "0           -                     -                          -               \n",
       "1           -                     -                          -               \n",
       "2           -                     -                          -               \n",
       "3           -                     -                          -               \n",
       "4           -                     -                          -               \n",
       "\n",
       "  Ontology Term Qualifier Names Ontology Term Types  \\\n",
       "0                -                        -           \n",
       "1                -                        -           \n",
       "2                -                        -           \n",
       "3                -                        -           \n",
       "4                -                        -           \n",
       "\n",
       "  Organism Name Interactor A Organism Name Interactor B  \n",
       "0        Homo sapiens               Homo sapiens         \n",
       "1        Homo sapiens               Homo sapiens         \n",
       "2        Homo sapiens               Homo sapiens         \n",
       "3        Homo sapiens               Homo sapiens         \n",
       "4        Homo sapiens               Homo sapiens         "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter for human-human interactions (taxonomy ID for humans is 9606) \n",
    "human_interactions = df[(df['Organism ID Interactor A'] == 9606) & (df['Organism ID Interactor B'] == 9606)] \n",
    "# Print the first 5 rows of the filtered DataFrame\n",
    "human_interactions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: center;\">\n",
       "      <th></th>\n",
       "      <th>#BioGRID Interaction ID</th>\n",
       "      <th>Entrez Gene Interactor A</th>\n",
       "      <th>Entrez Gene Interactor B</th>\n",
       "      <th>BioGRID ID Interactor A</th>\n",
       "      <th>BioGRID ID Interactor B</th>\n",
       "      <th>Systematic Name Interactor A</th>\n",
       "      <th>Systematic Name Interactor B</th>\n",
       "      <th>Official Symbol Interactor A</th>\n",
       "      <th>Official Symbol Interactor B</th>\n",
       "      <th>Synonyms Interactor A</th>\n",
       "      <th>Synonyms Interactor B</th>\n",
       "      <th>Experimental System</th>\n",
       "      <th>Experimental System Type</th>\n",
       "      <th>Author</th>\n",
       "      <th>Publication Source</th>\n",
       "      <th>Organism ID Interactor A</th>\n",
       "      <th>Organism ID Interactor B</th>\n",
       "      <th>Throughput</th>\n",
       "      <th>Score</th>\n",
       "      <th>Modification</th>\n",
       "      <th>Qualifications</th>\n",
       "      <th>Tags</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>SWISS-PROT Accessions Interactor A</th>\n",
       "      <th>TREMBL Accessions Interactor A</th>\n",
       "      <th>REFSEQ Accessions Interactor A</th>\n",
       "      <th>SWISS-PROT Accessions Interactor B</th>\n",
       "      <th>TREMBL Accessions Interactor B</th>\n",
       "      <th>REFSEQ Accessions Interactor B</th>\n",
       "      <th>Ontology Term IDs</th>\n",
       "      <th>Ontology Term Names</th>\n",
       "      <th>Ontology Term Categories</th>\n",
       "      <th>Ontology Term Qualifier IDs</th>\n",
       "      <th>Ontology Term Qualifier Names</th>\n",
       "      <th>Ontology Term Types</th>\n",
       "      <th>Organism Name Interactor A</th>\n",
       "      <th>Organism Name Interactor B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>103</td>\n",
       "      <td>6416</td>\n",
       "      <td>2318</td>\n",
       "      <td>112315</td>\n",
       "      <td>108607</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MAP2K4</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...</td>\n",
       "      <td>ABP-280|ABP280A|ABPA|ABPL|FLN2|MFM5|MPD4</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Marti A (1997)</td>\n",
       "      <td>PUBMED:9006895</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P45985</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_003001|NP_001268364</td>\n",
       "      <td>Q14315</td>\n",
       "      <td>Q59H94</td>\n",
       "      <td>NP_001120959|NP_001449</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>117</td>\n",
       "      <td>84665</td>\n",
       "      <td>88</td>\n",
       "      <td>124185</td>\n",
       "      <td>106603</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MYPN</td>\n",
       "      <td>ACTN2</td>\n",
       "      <td>CMD1DD|CMH22|MYOP|RCM4</td>\n",
       "      <td>CMD1AA</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Bang ML (2001)</td>\n",
       "      <td>PUBMED:11309420</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q86TC9</td>\n",
       "      <td>A0A087WX60</td>\n",
       "      <td>NP_001243197|NP_001243196|NP_115967</td>\n",
       "      <td>P35609</td>\n",
       "      <td>Q59FD9|F6THM6</td>\n",
       "      <td>NP_001094|NP_001265272|NP_001265273</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>183</td>\n",
       "      <td>90</td>\n",
       "      <td>2339</td>\n",
       "      <td>106605</td>\n",
       "      <td>108625</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>FNTA</td>\n",
       "      <td>ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI</td>\n",
       "      <td>FPTA|PGGT1A|PTAR2</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Wang T (1996)</td>\n",
       "      <td>PUBMED:8599089</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q04771</td>\n",
       "      <td>D3DPA4</td>\n",
       "      <td>NP_001104537|NP_001096</td>\n",
       "      <td>P49354</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_002018</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>278</td>\n",
       "      <td>2624</td>\n",
       "      <td>5371</td>\n",
       "      <td>108894</td>\n",
       "      <td>111384</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "      <td>DCML|IMD21|MONOMAC|NFE1B</td>\n",
       "      <td>MYL|PP8675|RNF71|TRIM19</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Tsuzuki S (2000)</td>\n",
       "      <td>PUBMED:10938104</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P23769</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_001139134|NP_116027|NP_001139133</td>\n",
       "      <td>P29590</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_150250|NP_150253|NP_150252|NP_150247|NP_150...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>418</td>\n",
       "      <td>6118</td>\n",
       "      <td>6774</td>\n",
       "      <td>112038</td>\n",
       "      <td>112651</td>\n",
       "      <td>RP4-547C9.3</td>\n",
       "      <td>-</td>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>REPA2|RP-A p32|RP-A p34|RPA32</td>\n",
       "      <td>ADMIO|APRF|HIES</td>\n",
       "      <td>Two-hybrid</td>\n",
       "      <td>physical</td>\n",
       "      <td>Kim J (2000)</td>\n",
       "      <td>PUBMED:10875894</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P15927</td>\n",
       "      <td>B4DUL2</td>\n",
       "      <td>NP_001342057|NP_002937|NP_001284487|NP_0013420...</td>\n",
       "      <td>P40763</td>\n",
       "      <td>-</td>\n",
       "      <td>NP_644805|NP_003141|NP_001356447|NP_001356443|...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #BioGRID Interaction ID Entrez Gene Interactor A Entrez Gene Interactor B  \\\n",
       "0            103                      6416                    2318             \n",
       "1            117                     84665                      88             \n",
       "2            183                        90                    2339             \n",
       "3            278                      2624                    5371             \n",
       "4            418                      6118                    6774             \n",
       "\n",
       "   BioGRID ID Interactor A  BioGRID ID Interactor B  \\\n",
       "0          112315                   108607            \n",
       "1          124185                   106603            \n",
       "2          106605                   108625            \n",
       "3          108894                   111384            \n",
       "4          112038                   112651            \n",
       "\n",
       "  Systematic Name Interactor A Systematic Name Interactor B  \\\n",
       "0                    -                       -                \n",
       "1                    -                       -                \n",
       "2                    -                       -                \n",
       "3                    -                       -                \n",
       "4          RP4-547C9.3                       -                \n",
       "\n",
       "  Official Symbol Interactor A Official Symbol Interactor B  \\\n",
       "0            MAP2K4                         FLNC              \n",
       "1              MYPN                        ACTN2              \n",
       "2             ACVR1                         FNTA              \n",
       "3             GATA2                          PML              \n",
       "4              RPA2                        STAT3              \n",
       "\n",
       "                Synonyms Interactor A                 \\\n",
       "0  JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...   \n",
       "1                             CMD1DD|CMH22|MYOP|RCM4   \n",
       "2            ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI   \n",
       "3                           DCML|IMD21|MONOMAC|NFE1B   \n",
       "4                      REPA2|RP-A p32|RP-A p34|RPA32   \n",
       "\n",
       "            Synonyms Interactor B           Experimental System  \\\n",
       "0  ABP-280|ABP280A|ABPA|ABPL|FLN2|MFM5|MPD4      Two-hybrid       \n",
       "1                                    CMD1AA      Two-hybrid       \n",
       "2                         FPTA|PGGT1A|PTAR2      Two-hybrid       \n",
       "3                   MYL|PP8675|RNF71|TRIM19      Two-hybrid       \n",
       "4                           ADMIO|APRF|HIES      Two-hybrid       \n",
       "\n",
       "  Experimental System Type       Author      Publication Source  \\\n",
       "0         physical            Marti A (1997)    PUBMED:9006895    \n",
       "1         physical            Bang ML (2001)   PUBMED:11309420    \n",
       "2         physical             Wang T (1996)    PUBMED:8599089    \n",
       "3         physical          Tsuzuki S (2000)   PUBMED:10938104    \n",
       "4         physical              Kim J (2000)   PUBMED:10875894    \n",
       "\n",
       "   Organism ID Interactor A  Organism ID Interactor B    Throughput   Score  \\\n",
       "0            9606                      9606            Low Throughput    -    \n",
       "1            9606                      9606            Low Throughput    -    \n",
       "2            9606                      9606            Low Throughput    -    \n",
       "3            9606                      9606            Low Throughput    -    \n",
       "4            9606                      9606            Low Throughput    -    \n",
       "\n",
       "  Modification Qualifications Tags Source Database  \\\n",
       "0       -             -         -       BIOGRID      \n",
       "1       -             -         -       BIOGRID      \n",
       "2       -             -         -       BIOGRID      \n",
       "3       -             -         -       BIOGRID      \n",
       "4       -             -         -       BIOGRID      \n",
       "\n",
       "  SWISS-PROT Accessions Interactor A TREMBL Accessions Interactor A  \\\n",
       "0               P45985                                  -             \n",
       "1               Q86TC9                         A0A087WX60             \n",
       "2               Q04771                             D3DPA4             \n",
       "3               P23769                                  -             \n",
       "4               P15927                             B4DUL2             \n",
       "\n",
       "            REFSEQ Accessions Interactor A            \\\n",
       "0                             NP_003001|NP_001268364   \n",
       "1                NP_001243197|NP_001243196|NP_115967   \n",
       "2                             NP_001104537|NP_001096   \n",
       "3                NP_001139134|NP_116027|NP_001139133   \n",
       "4  NP_001342057|NP_002937|NP_001284487|NP_0013420...   \n",
       "\n",
       "  SWISS-PROT Accessions Interactor B TREMBL Accessions Interactor B  \\\n",
       "0               Q14315                               Q59H94           \n",
       "1               P35609                        Q59FD9|F6THM6           \n",
       "2               P49354                                    -           \n",
       "3               P29590                                    -           \n",
       "4               P40763                                    -           \n",
       "\n",
       "            REFSEQ Accessions Interactor B           Ontology Term IDs  \\\n",
       "0                             NP_001120959|NP_001449          -          \n",
       "1                NP_001094|NP_001265272|NP_001265273          -          \n",
       "2                                          NP_002018          -          \n",
       "3  NP_150250|NP_150253|NP_150252|NP_150247|NP_150...          -          \n",
       "4  NP_644805|NP_003141|NP_001356447|NP_001356443|...          -          \n",
       "\n",
       "  Ontology Term Names Ontology Term Categories Ontology Term Qualifier IDs  \\\n",
       "0           -                     -                          -               \n",
       "1           -                     -                          -               \n",
       "2           -                     -                          -               \n",
       "3           -                     -                          -               \n",
       "4           -                     -                          -               \n",
       "\n",
       "  Ontology Term Qualifier Names Ontology Term Types  \\\n",
       "0                -                        -           \n",
       "1                -                        -           \n",
       "2                -                        -           \n",
       "3                -                        -           \n",
       "4                -                        -           \n",
       "\n",
       "  Organism Name Interactor A Organism Name Interactor B  \n",
       "0        Homo sapiens               Homo sapiens         \n",
       "1        Homo sapiens               Homo sapiens         \n",
       "2        Homo sapiens               Homo sapiens         \n",
       "3        Homo sapiens               Homo sapiens         \n",
       "4        Homo sapiens               Homo sapiens         "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter for human-human interactions\n",
    "human_physical_interactions = human_interactions[(human_interactions['Experimental System Type'] == 'physical')] \n",
    "# Print the first 5 rows of the filtered DataFrame\n",
    "human_physical_interactions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered DataFrame has been saved successfully!\n"
     ]
    }
   ],
   "source": [
    "# Save the filtered DataFrame to a new CSV file \n",
    "output_path = r'E:\\University\\BINM\\BIOGRID-Human-Human-Physical-Interactions.csv' \n",
    "human_physical_interactions.to_csv(output_path)\n",
    "print(\"Filtered DataFrame has been saved successfully!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "removing self loops and redundents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Redundant interactions and self-loops have been removed successfully!\n"
     ]
    }
   ],
   "source": [
    "# Remove self-loops\n",
    "human_physical_interactions = human_physical_interactions[\n",
    "    human_physical_interactions['Entrez Gene Interactor A'] != human_physical_interactions['Entrez Gene Interactor B']\n",
    "]\n",
    "\n",
    "# Convert the columns to the same type (string or integer)\n",
    "human_physical_interactions['Entrez Gene Interactor A'] = human_physical_interactions['Entrez Gene Interactor A'].astype(str)\n",
    "human_physical_interactions['Entrez Gene Interactor B'] = human_physical_interactions['Entrez Gene Interactor B'].astype(str)\n",
    "\n",
    "# Sort interactors to handle duplicates regardless of order\n",
    "# Create a new DataFrame with sorted interactors for consistent comparison\n",
    "sorted_interactions = human_physical_interactions.apply(\n",
    "    lambda x: pd.Series(sorted([x['Entrez Gene Interactor A'], x['Entrez Gene Interactor B']])), axis=1\n",
    ")\n",
    "\n",
    "# Rename the new columns for clarity\n",
    "sorted_interactions.columns = ['Interactor A', 'Interactor B']\n",
    "\n",
    "# Combine the sorted interactions with the original DataFrame to ensure we keep all columns\n",
    "human_physical_interactions['Interactor A'] = sorted_interactions['Interactor A']\n",
    "human_physical_interactions['Interactor B'] = sorted_interactions['Interactor B']\n",
    "\n",
    "# Drop duplicates based on the sorted interactors\n",
    "human_physical_interactions = human_physical_interactions.drop_duplicates(subset=['Interactor A', 'Interactor B'])\n",
    "\n",
    "#drop the added 'Interactor A' and 'Interactor B' columns if they are no longer needed\n",
    "human_physical_interactions = human_physical_interactions.drop(columns=['Interactor A', 'Interactor B'])\n",
    "\n",
    "# Save the cleaned DataFrame to a new file\n",
    "output_path_cleaned = r'E:\\University\\BINM\\BIOGRID-Human-Human-Physical-Interactions-Cleaned.csv'\n",
    "human_physical_interactions.to_csv(output_path_cleaned, index=False)\n",
    "\n",
    "print(\"Redundant interactions and self-loops have been removed successfully!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Syed Habibul Bashar\\AppData\\Local\\Temp\\ipykernel_15192\\491096485.py:3: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactome = pd.read_csv(file_path)\n"
     ]
    }
   ],
   "source": [
    "file_path = r'E:\\University\\BINM\\BIOGRID-Human-Human-Physical-Interactions-Cleaned.csv'\n",
    "\n",
    "interactome = pd.read_csv(file_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "LCC (largest connected componect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes in LCC: 19972\n",
      "Number of edges in LCC: 862217\n"
     ]
    }
   ],
   "source": [
    "# Create a graph from the interactome\n",
    "G = nx.Graph()\n",
    "for _, row in interactome.iterrows():\n",
    "    # Add edges between interacting proteins\n",
    "    G.add_edge(row['Official Symbol Interactor A'], row['Official Symbol Interactor B'])\n",
    "\n",
    "# Find the Largest Connected Component (LCC)\n",
    "largest_cc = max(nx.connected_components(G), key=len)\n",
    "interactome_lcc = G.subgraph(largest_cc)\n",
    "\n",
    "# Output basic statistics about the LCC\n",
    "print(f\"Number of nodes in LCC: {interactome_lcc.number_of_nodes()}\")\n",
    "print(f\"Number of edges in LCC: {interactome_lcc.number_of_edges()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Disease genes saved to 'all_interactome_genes.tsv'\n"
     ]
    }
   ],
   "source": [
    "interactome_nodes = pd.DataFrame(interactome_lcc.nodes)\n",
    "# Save the DataFrame to a TSV file \n",
    "outputpath = r'E:\\University\\BINM\\all_interactome_genes.tsv'\n",
    "interactome_nodes.to_csv(outputpath, sep='\\t', index=False) \n",
    "print(\"Disease genes saved to 'all_interactome_genes.tsv'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GDAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: center;\">\n",
       "      <th></th>\n",
       "      <th>DOClass</th>\n",
       "      <th>Disease</th>\n",
       "      <th>DiseaseClass</th>\n",
       "      <th>EvidenceIndexGDA</th>\n",
       "      <th>EvidenceLevelGDA</th>\n",
       "      <th>FirstRef</th>\n",
       "      <th>Gene</th>\n",
       "      <th>GeneDPI</th>\n",
       "      <th>GeneDSI</th>\n",
       "      <th>GeneFullName</th>\n",
       "      <th>GenePLI</th>\n",
       "      <th>HPOClass</th>\n",
       "      <th>LastRef</th>\n",
       "      <th>NumDiseasesAssociatedToGene</th>\n",
       "      <th>NumGenesDisease</th>\n",
       "      <th>NumPMIDs</th>\n",
       "      <th>NumVariantsAssociatedToGene</th>\n",
       "      <th>NumVariantsDisease</th>\n",
       "      <th>NumVariantsGDA</th>\n",
       "      <th>PathwayIDs</th>\n",
       "      <th>PathwayNames</th>\n",
       "      <th>ProteinClass</th>\n",
       "      <th>ScoreGDA</th>\n",
       "      <th>SemanticType</th>\n",
       "      <th>Type</th>\n",
       "      <th>UnitProt</th>\n",
       "      <th>diseaseUMLSCUI</th>\n",
       "      <th>geneEnsemblIDs</th>\n",
       "      <th>geneNcbiID</th>\n",
       "      <th>geneNcbiType</th>\n",
       "      <th>numCTs</th>\n",
       "      <th>numberChemsIncludedInEvidence</th>\n",
       "      <th>numberPmidsWithChemsIncludedInEvidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>disease of anatomical entity (7)</td>\n",
       "      <td>Osteoporoses</td>\n",
       "      <td>Nutritional and Metabolic Diseases,Musculoskel...</td>\n",
       "      <td>9.437.751.004.016.060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>PTH</td>\n",
       "      <td>1.31</td>\n",
       "      <td>0.35</td>\n",
       "      <td>parathyroid hormone</td>\n",
       "      <td>784.32.00</td>\n",
       "      <td>Abnormality of the musculoskeletal system (33127)</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>27</td>\n",
       "      <td>87</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>225</td>\n",
       "      <td>0</td>\n",
       "      <td>R-HSA-162582</td>\n",
       "      <td>Signal Transduction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Disease or Syndrome (T047)</td>\n",
       "      <td>disease</td>\n",
       "      <td>P01270</td>\n",
       "      <td>C0029456</td>\n",
       "      <td>ENSG00000152266</td>\n",
       "      <td>5741</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>80</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>disease of anatomical entity (7)</td>\n",
       "      <td>Osteoporoses</td>\n",
       "      <td>Nutritional and Metabolic Diseases,Musculoskel...</td>\n",
       "      <td>9.649.122.807.017.540</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>LRP5</td>\n",
       "      <td>1.31</td>\n",
       "      <td>0.46</td>\n",
       "      <td>LDL receptor related protein 5</td>\n",
       "      <td>1561.03.00</td>\n",
       "      <td>Abnormality of the musculoskeletal system (33127)</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>41</td>\n",
       "      <td>87</td>\n",
       "      <td>3</td>\n",
       "      <td>330</td>\n",
       "      <td>225</td>\n",
       "      <td>159</td>\n",
       "      <td>R-HSA-162582,R-HSA-1643685</td>\n",
       "      <td>Signal Transduction,Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Disease or Syndrome (T047)</td>\n",
       "      <td>disease</td>\n",
       "      <td>O75197</td>\n",
       "      <td>C0029456</td>\n",
       "      <td>ENSG00000162337</td>\n",
       "      <td>4041</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>disease of anatomical entity (7)</td>\n",
       "      <td>Osteoporoses</td>\n",
       "      <td>Nutritional and Metabolic Diseases,Musculoskel...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>WNT1</td>\n",
       "      <td>1.36</td>\n",
       "      <td>0.30</td>\n",
       "      <td>Wnt family member 1</td>\n",
       "      <td>452.27.00</td>\n",
       "      <td>Abnormality of the musculoskeletal system (33127)</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>8</td>\n",
       "      <td>87</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>225</td>\n",
       "      <td>2</td>\n",
       "      <td>R-HSA-1266738,R-HSA-162582</td>\n",
       "      <td>Developmental Biology,Signal Transduction</td>\n",
       "      <td>Signaling</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Disease or Syndrome (T047)</td>\n",
       "      <td>disease</td>\n",
       "      <td>P04628</td>\n",
       "      <td>C0029456</td>\n",
       "      <td>ENSG00000125084</td>\n",
       "      <td>7471</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>disease of anatomical entity (7)</td>\n",
       "      <td>Osteoporoses</td>\n",
       "      <td>Nutritional and Metabolic Diseases,Musculoskel...</td>\n",
       "      <td>9.464.285.714.285.710</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>TNFSF11</td>\n",
       "      <td>1.31</td>\n",
       "      <td>0.40</td>\n",
       "      <td>TNF superfamily member 11</td>\n",
       "      <td>1097.47.00</td>\n",
       "      <td>Abnormality of the musculoskeletal system (33127)</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>11</td>\n",
       "      <td>87</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>225</td>\n",
       "      <td>0</td>\n",
       "      <td>R-HSA-168256</td>\n",
       "      <td>Immune System</td>\n",
       "      <td>Signaling</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Disease or Syndrome (T047)</td>\n",
       "      <td>disease</td>\n",
       "      <td>Q5T9Y4,O14788,Q54A98</td>\n",
       "      <td>C0029456</td>\n",
       "      <td>ENSG00000120659</td>\n",
       "      <td>8600</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>disease of anatomical entity (7)</td>\n",
       "      <td>Osteoporoses</td>\n",
       "      <td>Nutritional and Metabolic Diseases,Musculoskel...</td>\n",
       "      <td>9.119.496.855.345.910</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>TNFRSF11B</td>\n",
       "      <td>1.31</td>\n",
       "      <td>0.38</td>\n",
       "      <td>TNF receptor superfamily member 11b</td>\n",
       "      <td>1469.04.00</td>\n",
       "      <td>Abnormality of the musculoskeletal system (33127)</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>13</td>\n",
       "      <td>87</td>\n",
       "      <td>1</td>\n",
       "      <td>49</td>\n",
       "      <td>225</td>\n",
       "      <td>0</td>\n",
       "      <td>R-HSA-168256</td>\n",
       "      <td>Immune System</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Disease or Syndrome (T047)</td>\n",
       "      <td>disease</td>\n",
       "      <td>O00300</td>\n",
       "      <td>C0029456</td>\n",
       "      <td>ENSG00000164761</td>\n",
       "      <td>4982</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               DOClass                 Disease     \\\n",
       "0  disease of anatomical entity (7)  Osteoporoses   \n",
       "1  disease of anatomical entity (7)  Osteoporoses   \n",
       "2  disease of anatomical entity (7)  Osteoporoses   \n",
       "3  disease of anatomical entity (7)  Osteoporoses   \n",
       "4  disease of anatomical entity (7)  Osteoporoses   \n",
       "\n",
       "                     DiseaseClass                       EvidenceIndexGDA     \\\n",
       "0  Nutritional and Metabolic Diseases,Musculoskel...  9.437.751.004.016.060   \n",
       "1  Nutritional and Metabolic Diseases,Musculoskel...  9.649.122.807.017.540   \n",
       "2  Nutritional and Metabolic Diseases,Musculoskel...                    1.0   \n",
       "3  Nutritional and Metabolic Diseases,Musculoskel...  9.464.285.714.285.710   \n",
       "4  Nutritional and Metabolic Diseases,Musculoskel...  9.119.496.855.345.910   \n",
       "\n",
       "   EvidenceLevelGDA  FirstRef    Gene     GeneDPI  GeneDSI  \\\n",
       "0        NaN          2005.0         PTH   1.31     0.35     \n",
       "1        NaN          2004.0        LRP5   1.31     0.46     \n",
       "2        NaN          2013.0        WNT1   1.36     0.30     \n",
       "3        NaN          2007.0     TNFSF11   1.31     0.40     \n",
       "4        NaN          2007.0   TNFRSF11B   1.31     0.38     \n",
       "\n",
       "              GeneFullName               GenePLI    \\\n",
       "0                  parathyroid hormone   784.32.00   \n",
       "1       LDL receptor related protein 5  1561.03.00   \n",
       "2                  Wnt family member 1   452.27.00   \n",
       "3            TNF superfamily member 11  1097.47.00   \n",
       "4  TNF receptor superfamily member 11b  1469.04.00   \n",
       "\n",
       "                       HPOClass                       LastRef  \\\n",
       "0  Abnormality of the musculoskeletal system (33127)  2019.0    \n",
       "1  Abnormality of the musculoskeletal system (33127)  2005.0    \n",
       "2  Abnormality of the musculoskeletal system (33127)  2013.0    \n",
       "3  Abnormality of the musculoskeletal system (33127)  2007.0    \n",
       "4  Abnormality of the musculoskeletal system (33127)  2007.0    \n",
       "\n",
       "   NumDiseasesAssociatedToGene  NumGenesDisease  NumPMIDs  \\\n",
       "0              27                     87             6      \n",
       "1              41                     87             3      \n",
       "2               8                     87             2      \n",
       "3              11                     87             1      \n",
       "4              13                     87             1      \n",
       "\n",
       "   NumVariantsAssociatedToGene  NumVariantsDisease  NumVariantsGDA  \\\n",
       "0               10                      225                 0        \n",
       "1              330                      225               159        \n",
       "2               28                      225                 2        \n",
       "3               50                      225                 0        \n",
       "4               49                      225                 0        \n",
       "\n",
       "           PathwayIDs                        PathwayNames                 \\\n",
       "0                R-HSA-162582                        Signal Transduction   \n",
       "1  R-HSA-162582,R-HSA-1643685                Signal Transduction,Disease   \n",
       "2  R-HSA-1266738,R-HSA-162582  Developmental Biology,Signal Transduction   \n",
       "3                R-HSA-168256                              Immune System   \n",
       "4                R-HSA-168256                              Immune System   \n",
       "\n",
       "  ProteinClass  ScoreGDA         SemanticType          Type    \\\n",
       "0         NaN      1.0    Disease or Syndrome (T047)  disease   \n",
       "1         NaN      1.0    Disease or Syndrome (T047)  disease   \n",
       "2   Signaling      1.0    Disease or Syndrome (T047)  disease   \n",
       "3   Signaling      1.0    Disease or Syndrome (T047)  disease   \n",
       "4         NaN      1.0    Disease or Syndrome (T047)  disease   \n",
       "\n",
       "         UnitProt       diseaseUMLSCUI  geneEnsemblIDs   geneNcbiID  \\\n",
       "0                P01270    C0029456     ENSG00000152266     5741      \n",
       "1                O75197    C0029456     ENSG00000162337     4041      \n",
       "2                P04628    C0029456     ENSG00000125084     7471      \n",
       "3  Q5T9Y4,O14788,Q54A98    C0029456     ENSG00000120659     8600      \n",
       "4                O00300    C0029456     ENSG00000164761     4982      \n",
       "\n",
       "    geneNcbiType   numCTs  numberChemsIncludedInEvidence  \\\n",
       "0  protein-coding    80                  2                 \n",
       "1  protein-coding     0                  0                 \n",
       "2  protein-coding     1                  0                 \n",
       "3  protein-coding    18                  0                 \n",
       "4  protein-coding    21                  0                 \n",
       "\n",
       "   numberPmidsWithChemsIncludedInEvidence  \n",
       "0                     2                    \n",
       "1                     0                    \n",
       "2                     0                    \n",
       "3                     0                    \n",
       "4                     0                    "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = r'E:\\University\\BINM\\Osteoporoses_DISEASES_Summary_GDA_CURATED_C0029456 - DISEASES_Summary_GDA_CURATED_C0029456.csv'\n",
    "\n",
    "dfo = pd.read_csv(file_path)\n",
    "\n",
    "dfo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Previous symbols</th>\n",
       "      <th>Alias symbols</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Accession numbers</th>\n",
       "      <th>RefSeq IDs</th>\n",
       "      <th>UniProt ID(supplied by UniProt)</th>\n",
       "      <th>NCBI Gene ID</th>\n",
       "      <th>Ensembl gene ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_130786</td>\n",
       "      <td>P04217</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NCRNA00181, A1BGAS, A1BG-AS</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>BC040926</td>\n",
       "      <td>NR_015380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>503538.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF, ASP, ACF64, ACF65, APOBEC1CF</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>AF271790</td>\n",
       "      <td>NM_014576</td>\n",
       "      <td>Q9NQ94</td>\n",
       "      <td>29974.0</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HGNC:6</td>\n",
       "      <td>A1S9T</td>\n",
       "      <td>symbol withdrawn, see [HGNC:12469](/data/gene-...</td>\n",
       "      <td>Symbol Withdrawn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>BX647329, X68728, M11313</td>\n",
       "      <td>NM_000014</td>\n",
       "      <td>P01023</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      HGNC ID Approved symbol  \\\n",
       "0      HGNC:5            A1BG   \n",
       "1  HGNC:37133        A1BG-AS1   \n",
       "2  HGNC:24086            A1CF   \n",
       "3      HGNC:6           A1S9T   \n",
       "4      HGNC:7             A2M   \n",
       "\n",
       "                                       Approved name            Status  \\\n",
       "0                             alpha-1-B glycoprotein          Approved   \n",
       "1                               A1BG antisense RNA 1          Approved   \n",
       "2                     APOBEC1 complementation factor          Approved   \n",
       "3  symbol withdrawn, see [HGNC:12469](/data/gene-...  Symbol Withdrawn   \n",
       "4                              alpha-2-macroglobulin          Approved   \n",
       "\n",
       "              Previous symbols                      Alias symbols Chromosome  \\\n",
       "0                          NaN                                NaN   19q13.43   \n",
       "1  NCRNA00181, A1BGAS, A1BG-AS                           FLJ23569   19q13.43   \n",
       "2                          NaN  ACF, ASP, ACF64, ACF65, APOBEC1CF   10q11.23   \n",
       "3                          NaN                                NaN        NaN   \n",
       "4                          NaN             FWP007, S863-7, CPAMD5   12p13.31   \n",
       "\n",
       "          Accession numbers RefSeq IDs UniProt ID(supplied by UniProt)  \\\n",
       "0                       NaN  NM_130786                          P04217   \n",
       "1                  BC040926  NR_015380                             NaN   \n",
       "2                  AF271790  NM_014576                          Q9NQ94   \n",
       "3                       NaN        NaN                             NaN   \n",
       "4  BX647329, X68728, M11313  NM_000014                          P01023   \n",
       "\n",
       "   NCBI Gene ID  Ensembl gene ID  \n",
       "0           1.0  ENSG00000121410  \n",
       "1      503538.0  ENSG00000268895  \n",
       "2       29974.0  ENSG00000148584  \n",
       "3           NaN              NaN  \n",
       "4           2.0  ENSG00000175899  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "file_path = r'E:\\University\\BINM\\hgnc_custom_set_2024-12-03.txt'\n",
    "\n",
    "dfh = pd.read_csv(file_path, delimiter='\\t')\n",
    "\n",
    "dfh.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DOClass                                   87\n",
       "Disease                                   87\n",
       "DiseaseClass                              87\n",
       "EvidenceIndexGDA                          81\n",
       "EvidenceLevelGDA                           0\n",
       "FirstRef                                  59\n",
       "Gene                                      87\n",
       "GeneDPI                                   87\n",
       "GeneDSI                                   87\n",
       "GeneFullName                              87\n",
       "GenePLI                                   80\n",
       "HPOClass                                  87\n",
       "LastRef                                   59\n",
       "NumDiseasesAssociatedToGene               87\n",
       "NumGenesDisease                           87\n",
       "NumPMIDs                                  87\n",
       "NumVariantsAssociatedToGene               87\n",
       "NumVariantsDisease                        87\n",
       "NumVariantsGDA                            87\n",
       "PathwayIDs                                78\n",
       "PathwayNames                              78\n",
       "ProteinClass                              50\n",
       "ScoreGDA                                  87\n",
       "SemanticType                              87\n",
       "Type                                      87\n",
       "UnitProt                                  84\n",
       "diseaseUMLSCUI                            87\n",
       "geneEnsemblIDs                            85\n",
       "geneNcbiID                                87\n",
       "geneNcbiType                              87\n",
       "numCTs                                    87\n",
       "numberChemsIncludedInEvidence             87\n",
       "numberPmidsWithChemsIncludedInEvidence    87\n",
       "HGNC ID                                   85\n",
       "Approved symbol                           85\n",
       "Approved name                             85\n",
       "Status                                    85\n",
       "Previous symbols                          35\n",
       "Alias symbols                             69\n",
       "Chromosome                                85\n",
       "Accession numbers                         70\n",
       "RefSeq IDs                                65\n",
       "UniProt ID(supplied by UniProt)           84\n",
       "NCBI Gene ID                              85\n",
       "Ensembl gene ID                           85\n",
       "dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge the data frames on the gene columns \n",
    "merged_df = dfo.merge(dfh, left_on='Gene', right_on='Approved symbol', how='left', suffixes=('_dfo', '_dfh'))\n",
    "merged_df.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Inconsistencies found:\n",
      "        Gene      geneEnsemblIDs  Ensembl gene ID   geneNcbiID  NCBI Gene ID  \\\n",
      "12         CALCA  ENSG00000110680  ENSG00000110680        796.0     796.0      \n",
      "84  LOC130059892                               NaN  130059892.0       NaN      \n",
      "86  LOC126862586                               NaN  126862586.0       NaN      \n",
      "\n",
      "      UnitProt    UniProt ID(supplied by UniProt)  \n",
      "12  P01258,P06881          P01258, P06881          \n",
      "84                                    NaN          \n",
      "86                                    NaN          \n"
     ]
    }
   ],
   "source": [
    "# Convert relevant columns to the same data type (string)\n",
    "dfo['geneEnsemblIDs'] = dfo['geneEnsemblIDs'].astype(str)\n",
    "dfo['geneNcbiID'] = dfo['geneNcbiID'].astype(str)  # Format to 1 decimal place\n",
    "dfo['UnitProt'] = dfo['UnitProt'].astype(str)\n",
    "\n",
    "dfh['Ensembl gene ID'] = dfh['Ensembl gene ID'].astype(str)\n",
    "dfh['NCBI Gene ID'] = dfh['NCBI Gene ID'].astype(str) # Format to 1 decimal place\n",
    "dfh['UniProt ID(supplied by UniProt)'] = dfh['UniProt ID(supplied by UniProt)'].astype(str)\n",
    "\n",
    "# Merge the data frames on the gene columns\n",
    "merged_df = dfo.merge(dfh, left_on='Gene', right_on='Approved symbol', how='left', suffixes=('_dfo', '_dfh'))\n",
    "\n",
    "# Function to check if there's at least one matching UniProt ID\n",
    "def check_unitprot_match(row):\n",
    "    dfo_ids = set(row['UnitProt'].split(','))\n",
    "    dfh_id = row['UniProt ID(supplied by UniProt)']\n",
    "    return dfh_id in dfo_ids\n",
    "\n",
    "# Check for inconsistencies in the specified columns\n",
    "inconsistencies = merged_df[(merged_df['geneEnsemblIDs'] != merged_df['Ensembl gene ID']) |\n",
    "                            (merged_df['geneNcbiID'] != merged_df['NCBI Gene ID']) |\n",
    "                            (~merged_df.apply(check_unitprot_match, axis=1))]\n",
    "\n",
    "# Print the inconsistencies\n",
    "print(\"Inconsistencies found:\")\n",
    "print(inconsistencies[['Gene', 'geneEnsemblIDs', 'Ensembl gene ID', \n",
    "                       'geneNcbiID', 'NCBI Gene ID', 'UnitProt', 'UniProt ID(supplied by UniProt)']])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1.3 finding genes in LCC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of disease genes in the LCC: 84\n",
      "Disease genes in the LCC:\n",
      "{'REN', 'CCT2', 'LEPR', 'GSN', 'OXCT1', 'KL', 'PKM', 'ANXA2', 'SOD2', 'TNFSF11', 'TNFRSF1B', 'BGLAP', 'LTF', 'TNFRSF11A', 'ANTXR2', 'P4HB', 'GPC6', 'CALCR', 'TUBA1C', 'IGF1R', 'CYP17A1', 'RYR1', 'CA2', 'GAPDH', 'MGLL', 'BMP4', 'ADCY5', 'CYP24A1', 'CAP1', 'TLN1', 'PLEK', 'PDLIM4', 'OPA1', 'IL6R', 'IL32', 'FGA', 'COL1A1', 'RAB7B', 'FGB', 'BMP2', 'RSU1', 'TGFB1', 'PGLS', 'PNP', 'DBP', 'GPX1', 'TPI1', 'COL1A2', 'PSMA5', 'LRP5', 'PSMA2', 'DAAM2', 'ESR2', 'ENO1', 'VCL', 'GSTP1', 'CXCR4', 'TUBA1B', 'PRDX3', 'IL6', 'WNT1', 'CYP19A1', 'GPD2', 'TNFRSF11B', 'ACTG1', 'TPM4', 'SERPINF1', 'PARK7', 'GORAB', 'ZDHHC13', 'IL1RN', 'POMC', 'ATIC', 'NRIP1', 'ESR1', 'WDR1', 'CALCA', 'SRC', 'LEP', 'PTH', 'IDH2', 'U2AF1', 'IL1A', 'CLEC11A'}\n"
     ]
    }
   ],
   "source": [
    "# Load disease-gene associations\n",
    "disease_genes = dfo\n",
    "disease_gene_set = set(disease_genes['Gene'])  # Extract disease genes into a set for quick lookup\n",
    "\n",
    "# Extract nodes from the LCC\n",
    "lcc_nodes = set(interactome_lcc.nodes)\n",
    "\n",
    "# Find disease genes present in the LCC\n",
    "disease_genes_in_lcc = lcc_nodes.intersection(disease_gene_set)\n",
    "\n",
    "# Output results\n",
    "print(f\"Number of disease genes in the LCC: {len(disease_genes_in_lcc)}\")\n",
    "print(\"Disease genes in the LCC:\")\n",
    "print(disease_genes_in_lcc)\n",
    "\n",
    "#Save the list of disease genes in the LCC\n",
    "pd.DataFrame({'Disease Genes in LCC': list(disease_genes_in_lcc)}).to_csv('disease_genes_in_lcc.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Disease genes in the LCC have been saved as a TSV file: 'disease_genes_in_lcc.tsv'\n"
     ]
    }
   ],
   "source": [
    "# Save the list of disease genes in the LCC as a TSV\n",
    "pd.DataFrame({'Disease Genes in LCC': list(disease_genes_in_lcc)}).to_csv('disease_genes_in_lcc.tsv', sep=\"\\t\", index=False)\n",
    "print(\"Disease genes in the LCC have been saved as a TSV file: 'disease_genes_in_lcc.tsv'\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Desease LCC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes in the Disease LCC: 47\n",
      "Number of edges in the Disease LCC: 113\n"
     ]
    }
   ],
   "source": [
    "# Filter edges in the LCC for disease genes\n",
    "disease_interactions = [\n",
    "    (u, v) for u, v in interactome_lcc.edges\n",
    "    if u in disease_genes_in_lcc and v in disease_genes_in_lcc\n",
    "]\n",
    "\n",
    "# Create a graph for the disease interactome\n",
    "G_disease = nx.Graph()\n",
    "G_disease.add_edges_from(disease_interactions)\n",
    "\n",
    "# Find the Largest Connected Component (LCC) of the disease interactome\n",
    "largest_cc_disease = max(nx.connected_components(G_disease), key=len)\n",
    "disease_lcc = G_disease.subgraph(largest_cc_disease)\n",
    "\n",
    "# Output basic statistics\n",
    "print(f\"Number of nodes in the Disease LCC: {disease_lcc.number_of_nodes()}\")\n",
    "print(f\"Number of edges in the Disease LCC: {disease_lcc.number_of_edges()}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Node Degree  Betweenness Centrality  Eigenvector Centrality  \\\n",
      "count   47.000000          47.000000               47.000000         \n",
      "mean     4.808511           0.038175                0.112414         \n",
      "std      4.100044           0.067539                0.093954         \n",
      "min      1.000000           0.000000                0.002788         \n",
      "25%      2.000000           0.000000                0.043515         \n",
      "50%      4.000000           0.009700                0.097069         \n",
      "75%      6.000000           0.043816                0.149719         \n",
      "max     22.000000           0.350636                0.465064         \n",
      "\n",
      "       Closeness Centrality  Betweenness/Degree  \n",
      "count        47.000000            47.000000      \n",
      "mean          0.381920             0.005007      \n",
      "std           0.072194             0.006116      \n",
      "min           0.237113             0.000000      \n",
      "25%           0.343303             0.000000      \n",
      "50%           0.380165             0.002260      \n",
      "75%           0.433962             0.009234      \n",
      "max           0.567901             0.021739      \n"
     ]
    }
   ],
   "source": [
    "# Compute node degree\n",
    "node_degree = dict(disease_lcc.degree())\n",
    "\n",
    "# Compute betweenness centrality\n",
    "betweenness_centrality = nx.betweenness_centrality(disease_lcc)\n",
    "\n",
    "# Compute eigenvector centrality\n",
    "eigenvector_centrality = nx.eigenvector_centrality(disease_lcc)\n",
    "\n",
    "# Compute closeness centrality\n",
    "closeness_centrality = nx.closeness_centrality(disease_lcc)\n",
    "\n",
    "# Avoid dividing by zero for nodes with zero degree\n",
    "ratio_betweenness_degree = {\n",
    "    node: betweenness_centrality[node] / degree if degree > 0 else 0\n",
    "    for node, degree in node_degree.items()\n",
    "}\n",
    "\n",
    "# Combine all metrics into a DataFrame for easy analysis\n",
    "metrics_df = pd.DataFrame({\n",
    "    'Node': list(disease_lcc.nodes),\n",
    "    'Node Degree': [node_degree[node] for node in disease_lcc.nodes],\n",
    "    'Betweenness Centrality': [betweenness_centrality[node] for node in disease_lcc.nodes],\n",
    "    'Eigenvector Centrality': [eigenvector_centrality[node] for node in disease_lcc.nodes],\n",
    "    'Closeness Centrality': [closeness_centrality[node] for node in disease_lcc.nodes],\n",
    "    'Betweenness/Degree': [ratio_betweenness_degree[node] for node in disease_lcc.nodes]\n",
    "})\n",
    "\n",
    "print(metrics_df.describe())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      Node     Node Degree  Betweenness Centrality  Eigenvector Centrality  \\\n",
      "8        ESR2      22              0.350636                0.465064          \n",
      "4        ESR1      13              0.128508                0.304440          \n",
      "6         SRC      13              0.181352                0.314973          \n",
      "24       ENO1      12              0.224414                0.256062          \n",
      "12      GAPDH      11              0.100957                0.237012          \n",
      "34      PARK7      10              0.065251                0.137031          \n",
      "36       TPI1       8              0.086163                0.153290          \n",
      "13       CCT2       7              0.032903                0.191723          \n",
      "19      ACTG1       7              0.042035                0.169749          \n",
      "2        CAP1       6              0.107311                0.141101          \n",
      "25        VCL       6              0.025448                0.194869          \n",
      "7        P4HB       6              0.079007                0.157919          \n",
      "41        PKM       6              0.025105                0.116382          \n",
      "43       IDH2       6              0.005489                0.113810          \n",
      "15       TLN1       6              0.010990                0.209500          \n",
      "22       TPM4       5              0.016212                0.146148          \n",
      "28      GSTP1       5              0.020194                0.146114          \n",
      "16     TUBA1B       5              0.046452                0.121793          \n",
      "10        GSN       5              0.003572                0.188162          \n",
      "14     TUBA1C       4              0.003612                0.111016          \n",
      "42       SOD2       4              0.000919                0.067590          \n",
      "29       RYR1       4              0.000719                0.104767          \n",
      "3       IGF1R       4              0.045217                0.102666          \n",
      "0      COL1A1       4              0.043478                0.021472          \n",
      "40      OXCT1       3              0.044155                0.028030          \n",
      "27      PSMA5       3              0.006780                0.084771          \n",
      "37      PRDX3       3              0.013706                0.047217          \n",
      "17      GORAB       3              0.000000                0.045430          \n",
      "26       OPA1       3              0.009700                0.085496          \n",
      "21        FGB       3              0.000621                0.097069          \n",
      "18      TGFB1       3              0.001852                0.067442          \n",
      "9       NRIP1       3              0.000805                0.118896          \n",
      "44        PNP       3              0.000000                0.038211          \n",
      "35       WDR1       3              0.027159                0.058931          \n",
      "23    TNFSF11       2              0.043478                0.041600          \n",
      "30      PSMA2       2              0.000000                0.071395          \n",
      "20        FGA       2              0.000000                0.072992          \n",
      "11      ANXA2       2              0.000000                0.080429          \n",
      "33       PGLS       1              0.000000                0.017794          \n",
      "32      CXCR4       1              0.000000                0.020505          \n",
      "31       GPC6       1              0.000000                0.060388          \n",
      "38      U2AF1       1              0.000000                0.015815          \n",
      "39       ATIC       1              0.000000                0.003640          \n",
      "1      COL1A2       1              0.000000                0.002788          \n",
      "5     CYP17A1       1              0.000000                0.013331          \n",
      "45        CA2       1              0.000000                0.033250          \n",
      "46  TNFRSF11B       1              0.000000                0.005402          \n",
      "\n",
      "    Closeness Centrality  Betweenness/Degree  \n",
      "8         0.567901             0.015938       \n",
      "4         0.484211             0.009885       \n",
      "6         0.500000             0.013950       \n",
      "24        0.522727             0.018701       \n",
      "12        0.479167             0.009178       \n",
      "34        0.400000             0.006525       \n",
      "36        0.425926             0.010770       \n",
      "13        0.442308             0.004700       \n",
      "19        0.438095             0.006005       \n",
      "2         0.433962             0.017885       \n",
      "25        0.455446             0.004241       \n",
      "7         0.464646             0.013168       \n",
      "41        0.403509             0.004184       \n",
      "43        0.380165             0.000915       \n",
      "15        0.433962             0.001832       \n",
      "22        0.414414             0.003242       \n",
      "28        0.438095             0.004039       \n",
      "16        0.377049             0.009290       \n",
      "10        0.433962             0.000714       \n",
      "14        0.393162             0.000903       \n",
      "42        0.321678             0.000230       \n",
      "29        0.389831             0.000180       \n",
      "3         0.373984             0.011304       \n",
      "0         0.308725             0.010870       \n",
      "40        0.312925             0.014718       \n",
      "27        0.393162             0.002260       \n",
      "37        0.345865             0.004569       \n",
      "17        0.328571             0.000000       \n",
      "26        0.383333             0.003233       \n",
      "21        0.373984             0.000207       \n",
      "18        0.362205             0.000617       \n",
      "9         0.396552             0.000268       \n",
      "44        0.345865             0.000000       \n",
      "35        0.383333             0.009053       \n",
      "23        0.340741             0.021739       \n",
      "30        0.373984             0.000000       \n",
      "20        0.368000             0.000000       \n",
      "11        0.359375             0.000000       \n",
      "33        0.287500             0.000000       \n",
      "32        0.319444             0.000000       \n",
      "31        0.365079             0.000000       \n",
      "38        0.275449             0.000000       \n",
      "39        0.239583             0.000000       \n",
      "1         0.237113             0.000000       \n",
      "5         0.273810             0.000000       \n",
      "45        0.345865             0.000000       \n",
      "46        0.255556             0.000000       \n"
     ]
    }
   ],
   "source": [
    "# Sort the DataFrame by Node Degree in descending order\n",
    "metrics_df_sorted = metrics_df.sort_values(by='Node Degree', ascending=False)\n",
    "\n",
    "# Select the first 47 nodes\n",
    "top_47_metrics = metrics_df_sorted.head(47)\n",
    "\n",
    "# Save the top 47 metrics to a CSV file\n",
    "top_47_metrics.to_csv('top_47_disease_lcc_metrics.csv', index=False)\n",
    "\n",
    "# Display the table\n",
    "print(top_47_metrics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table saved as 'top_47_disease_lcc_metrics.csv'\n"
     ]
    }
   ],
   "source": [
    "# Save the table to a CSV file\n",
    "output_path = r'E:\\University\\BINM\\top_47_disease_lcc_metrics.csv'\n",
    "top_47_metrics.to_csv(output_path, index=False)\n",
    "\n",
    "print(\"Table saved as 'top_47_disease_lcc_metrics.csv'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Comparative analysis of the disease genes identification algorithms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Disease genes saved to '84_disease_genes.tsv'\n"
     ]
    }
   ],
   "source": [
    "disease_genes_1c = pd.DataFrame(disease_lcc, columns=['Gene'])\n",
    "# Save the DataFrame to a TSV file \n",
    "outputpath = r'E:\\University\\BINM\\84_disease_genes.tsv'\n",
    "disease_genes_1c.to_csv(outputpath, sep='\\t', index=False) \n",
    "print(\"Disease genes saved to '84_disease_genes.tsv'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Disease genes saved to 'interactome_lcc_edgelist.tsv'\n"
     ]
    }
   ],
   "source": [
    "disease_edge_list = pd.DataFrame(interactome_lcc.edges)\n",
    "outputpath = r'E:\\University\\BINM\\interactome_lcc_edgelist.tsv'\n",
    "disease_edge_list.to_csv(outputpath, sep='\\t', index=False) \n",
    "print(\"Disease genes saved to 'interactome_lcc_edgelist.tsv'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E:\\University\\BINM\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Syed Habibul Bashar\\AppData\\Roaming\\Python\\Python39\\site-packages\\IPython\\core\\magics\\osm.py:417: UserWarning: using dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    }
   ],
   "source": [
    "cd E:\\University\\BINM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_graph_and_genes(edgelist_file, seed_genes_file):\n",
    "    \"\"\"\n",
    "    Load the graph and seed genes.\n",
    "\n",
    "    Args:\n",
    "        edgelist_file (str): Path to the edgelist file.\n",
    "        seed_genes_file (str): Path to the file containing seed genes.\n",
    "\n",
    "    Returns:\n",
    "        tuple: Graph object and list of valid seed genes.\n",
    "    \"\"\"\n",
    "    # Load the graph\n",
    "    G = nx.read_edgelist(edgelist_file, delimiter=\"\\t\")\n",
    "\n",
    "    # Read the seed genes TSV file\n",
    "    seed_genes_df = pd.read_csv(seed_genes_file, delimiter=\"\\t\")\n",
    "    seed_genes = seed_genes_df['Disease Genes in LCC'].tolist()\n",
    "\n",
    "    # Validate seed genes\n",
    "    valid_seed_genes = [gene for gene in seed_genes if gene in G.nodes()]\n",
    "    return G, valid_seed_genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "def diffusion_ranking_sparse_exclude_seeds(G, seed_genes, t):\n",
    "    \"\"\"\n",
    "    Perform diffusion-based ranking using sparse matrices and exclude seed genes.\n",
    "\n",
    "    Args:\n",
    "        G (networkx.Graph): Input graph.\n",
    "        seed_genes (list): List of seed genes.\n",
    "        t (float): Diffusion time.\n",
    "\n",
    "    Returns:\n",
    "        list: Ranked list of nodes based on diffusion scores, excluding seed genes.\n",
    "    \"\"\"\n",
    "    nodes = list(G.nodes())\n",
    "    node_to_idx = {node: idx for idx, node in enumerate(nodes)}\n",
    "    A = nx.adjacency_matrix(G, nodelist=nodes)\n",
    "    L = csgraph.laplacian(A, normed=False)\n",
    "\n",
    "    H0 = np.zeros(len(nodes))\n",
    "    for gene in seed_genes:\n",
    "        if gene in node_to_idx:\n",
    "            H0[node_to_idx[gene]] = 1\n",
    "\n",
    "    Ht = expm_multiply(-t * L, H0)\n",
    "    ranked_nodes = [nodes[idx] for idx in np.argsort(-Ht)]\n",
    "    ranked_nodes = [node for node in ranked_nodes if node not in seed_genes]\n",
    "    return ranked_nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "def diamond_ranking(G, seed_genes, max_iterations=50):\n",
    "    \"\"\"\n",
    "    Perform DIAMOnD algorithm to expand the seed genes.\n",
    "\n",
    "    Args:\n",
    "        G (networkx.Graph): Input graph.\n",
    "        seed_genes (list): List of seed genes.\n",
    "        max_iterations (int): Number of nodes to add.\n",
    "\n",
    "    Returns:\n",
    "        list: Ranked list of additional nodes based on significance.\n",
    "    \"\"\"\n",
    "    added_nodes = []\n",
    "    seed_set = set(seed_genes)\n",
    "\n",
    "    for _ in range(max_iterations):\n",
    "        candidate_nodes = [node for node in G.nodes if node not in seed_set]\n",
    "        node_pvalues = {}\n",
    "\n",
    "        for node in candidate_nodes:\n",
    "            neighbors = set(G.neighbors(node))\n",
    "            overlap = len(neighbors & seed_set)\n",
    "            M = len(G.nodes)\n",
    "            n = len(seed_set)\n",
    "            N = len(neighbors)\n",
    "            x = overlap\n",
    "            p_value = hypergeom.sf(x - 1, M, n, N)\n",
    "            node_pvalues[node] = p_value\n",
    "\n",
    "        if not node_pvalues:\n",
    "            break\n",
    "        next_node = min(node_pvalues, key=node_pvalues.get)\n",
    "        seed_set.add(next_node)\n",
    "        added_nodes.append(next_node)\n",
    "\n",
    "    return added_nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [],
   "source": [
    "def diable_ranking(G, seed_genes, max_iterations=50):\n",
    "    \"\"\"\n",
    "    Perform DiaBLE algorithm with adjusted universe expansion.\n",
    "\n",
    "    Args:\n",
    "        G (networkx.Graph): Input graph.\n",
    "        seed_genes (list): List of seed genes.\n",
    "        max_iterations (int): Number of iterations.\n",
    "\n",
    "    Returns:\n",
    "        list: Ranked list of additional nodes based on significance.\n",
    "    \"\"\"\n",
    "    added_nodes = []\n",
    "    seed_set = set(seed_genes)\n",
    "\n",
    "    for _ in range(max_iterations):\n",
    "        candidate_nodes = set()\n",
    "        for seed in seed_set:\n",
    "            candidate_nodes.update(G.neighbors(seed))\n",
    "\n",
    "        candidate_nodes = [node for node in candidate_nodes if node not in seed_set]\n",
    "        node_pvalues = {}\n",
    "\n",
    "        for node in candidate_nodes:\n",
    "            neighbors = set(G.neighbors(node))\n",
    "            overlap = len(neighbors & seed_set)\n",
    "            M = len(G.nodes)\n",
    "            n = len(seed_set)\n",
    "            N = len(neighbors)\n",
    "            x = overlap\n",
    "            p_value = hypergeom.sf(x - 1, M, n, N)\n",
    "            node_pvalues[node] = p_value\n",
    "\n",
    "        if not node_pvalues:\n",
    "            break\n",
    "        next_node = min(node_pvalues, key=node_pvalues.get)\n",
    "        seed_set.add(next_node)\n",
    "        added_nodes.append(next_node)\n",
    "\n",
    "    return added_nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate_performance(predicted_genes, true_genes, top_k):\n",
    "    \"\"\"\n",
    "    Evaluate precision, recall, and F1-score at top-k predictions.\n",
    "\n",
    "    Args:\n",
    "        predicted_genes (list): Ranked list of predicted genes.\n",
    "        true_genes (set): Set of true disease genes.\n",
    "        top_k (int): Number of top predictions to evaluate.\n",
    "\n",
    "    Returns:\n",
    "        dict: Precision, Recall, and F1-score.\n",
    "    \"\"\"\n",
    "    predicted_set = set(predicted_genes[:top_k])\n",
    "    true_positives = len(predicted_set & true_genes)\n",
    "    false_positives = len(predicted_set - true_genes)\n",
    "    false_negatives = len(true_genes - predicted_set)\n",
    "\n",
    "    precision = true_positives / (true_positives + false_positives) if (true_positives + false_positives) > 0 else 0\n",
    "    recall = true_positives / (true_positives + false_negatives) if (true_positives + false_negatives) > 0 else 0\n",
    "    f1 = 2 * precision * recall / (precision + recall) if (precision + recall) > 0 else 0\n",
    "\n",
    "    return {\"precision\": precision, \"recall\": recall, \"f1\": f1}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cross_validate_ranking(G, seed_genes, true_genes, algorithm, params, top_k, folds=5):\n",
    "    \"\"\"\n",
    "    Perform cross-validation for a ranking algorithm.\n",
    "\n",
    "    Args:\n",
    "        G (networkx.Graph): Input graph.\n",
    "        seed_genes (list): List of seed genes.\n",
    "        true_genes (set): Set of true disease genes.\n",
    "        algorithm (callable): Ranking algorithm to evaluate.\n",
    "        params (dict): Parameters for the ranking algorithm.\n",
    "        top_k (int): Number of top predictions to evaluate.\n",
    "        folds (int): Number of cross-validation folds.\n",
    "\n",
    "    Returns:\n",
    "        dict: Average and standard deviation of precision, recall, and F1-score.\n",
    "    \"\"\"\n",
    "    seed_genes = list(seed_genes)\n",
    "    np.random.shuffle(seed_genes)\n",
    "    fold_size = len(seed_genes) // folds\n",
    "    precision_scores, recall_scores, f1_scores = [], [], []\n",
    "\n",
    "    for i in range(folds):\n",
    "        test_set = set(seed_genes[i * fold_size:(i + 1) * fold_size])\n",
    "        train_set = set(seed_genes) - test_set\n",
    "\n",
    "        predicted_genes = algorithm(G, list(train_set), **params)\n",
    "        metrics = evaluate_performance(predicted_genes, test_set, top_k)\n",
    "        precision_scores.append(metrics[\"precision\"])\n",
    "        recall_scores.append(metrics[\"recall\"])\n",
    "        f1_scores.append(metrics[\"f1\"])\n",
    "\n",
    "    return {\n",
    "        \"precision\": (np.mean(precision_scores), np.std(precision_scores)),\n",
    "        \"recall\": (np.mean(recall_scores), np.std(recall_scores)),\n",
    "        \"f1\": (np.mean(f1_scores), np.std(f1_scores)),\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_algorithms(G, valid_seed_genes, true_genes, top_k, folds=5):\n",
    "    algorithms = {\n",
    "        \"DIAMOnD\": {\n",
    "            \"func\": diamond_ranking,\n",
    "            \"params\": {\"max_iterations\": top_k},\n",
    "        },\n",
    "        \"DiaBLE\": {\n",
    "            \"func\": diable_ranking,\n",
    "            \"params\": {\"max_iterations\": top_k},\n",
    "        },\n",
    "        \"Diffusion_t=0.002\": {\n",
    "            \"func\": diffusion_ranking_sparse_exclude_seeds,\n",
    "            \"params\": {\"t\": 0.002},\n",
    "        },\n",
    "        \"Diffusion_t=0.005\": {\n",
    "            \"func\": diffusion_ranking_sparse_exclude_seeds,\n",
    "            \"params\": {\"t\": 0.005},\n",
    "        },\n",
    "        \"Diffusion_t=0.01\": {\n",
    "            \"func\": diffusion_ranking_sparse_exclude_seeds,\n",
    "            \"params\": {\"t\": 0.01},\n",
    "        },\n",
    "    }\n",
    "\n",
    "    results = {}\n",
    "    for name, algorithm in algorithms.items():\n",
    "        print(f\"Running cross-validation for {name}...\")\n",
    "        results[name] = cross_validate_ranking(\n",
    "            G,\n",
    "            valid_seed_genes,\n",
    "            true_genes,\n",
    "            algorithm[\"func\"],\n",
    "            algorithm[\"params\"],\n",
    "            top_k,\n",
    "            folds,\n",
    "        )\n",
    "\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing DIAMOnD, DiaBLE, and Diffusion-based algorithms...\n",
      "Running cross-validation for DIAMOnD...\n",
      "Running cross-validation for DiaBLE...\n",
      "Running cross-validation for Diffusion_t=0.002...\n",
      "Running cross-validation for Diffusion_t=0.005...\n",
      "Running cross-validation for Diffusion_t=0.01...\n",
      "\n",
      "Algorithm: DIAMOnD\n",
      "Precision: 0.0200  0.0126\n",
      "Recall: 0.0625  0.0395\n",
      "F1-score: 0.0303  0.0192\n",
      "\n",
      "Algorithm: DiaBLE\n",
      "Precision: 0.0160  0.0196\n",
      "Recall: 0.0500  0.0612\n",
      "F1-score: 0.0242  0.0297\n",
      "\n",
      "Algorithm: Diffusion_t=0.002\n",
      "Precision: 0.0080  0.0098\n",
      "Recall: 0.0250  0.0306\n",
      "F1-score: 0.0121  0.0148\n",
      "\n",
      "Algorithm: Diffusion_t=0.005\n",
      "Precision: 0.0080  0.0098\n",
      "Recall: 0.0250  0.0306\n",
      "F1-score: 0.0121  0.0148\n",
      "\n",
      "Algorithm: Diffusion_t=0.01\n",
      "Precision: 0.0000  0.0000\n",
      "Recall: 0.0000  0.0000\n",
      "F1-score: 0.0000  0.0000\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # Load the graph and seed genes\n",
    "    edgelist_file = \"interactome_lcc_edgelist.tsv\"\n",
    "    seed_genes_file = \"84_disease_genes.tsv\"\n",
    "    G, valid_seed_genes = load_graph_and_genes(edgelist_file, seed_genes_file)\n",
    "    true_genes = set(valid_seed_genes)\n",
    "\n",
    "    # Evaluation parameters\n",
    "    top_k = 50  # Evaluate top 50 predictions\n",
    "    folds = 5\n",
    "\n",
    "    # Compare all algorithms\n",
    "    print(\"Comparing DIAMOnD, DiaBLE, and Diffusion-based algorithms...\")\n",
    "    results = compare_algorithms(G, valid_seed_genes, true_genes, top_k, folds)\n",
    "\n",
    "    # Output results\n",
    "    for algorithm, metrics in results.items():\n",
    "        print(f\"\\nAlgorithm: {algorithm}\")\n",
    "        print(f\"Precision: {metrics['precision'][0]:.4f}  {metrics['precision'][1]:.4f}\")\n",
    "        print(f\"Recall: {metrics['recall'][0]:.4f}  {metrics['recall'][1]:.4f}\")\n",
    "        print(f\"F1-score: {metrics['f1'][0]:.4f}  {metrics['f1'][1]:.4f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.Putative disease gene identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DIAMOnD algorithm...\n",
      "Top 100 putative disease genes saved to 'top_100_putative_disease_genes_diamond.csv'\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # Load the graph and seed genes\n",
    "    edgelist_file = \"interactome_lcc_edgelist.tsv\"\n",
    "    seed_genes_file = \"84_disease_genes.tsv\"\n",
    "    G, valid_seed_genes = load_graph_and_genes(edgelist_file, seed_genes_file)\n",
    "\n",
    "    # DIAMOnD parameters\n",
    "    max_iterations = 100  # We need the top 100 putative disease genes\n",
    "\n",
    "    print(\"Running DIAMOnD algorithm...\")\n",
    "\n",
    "    # Run the DIAMOnD algorithm with all known GDAs as seed genes\n",
    "    predicted_genes = diamond_ranking(G, valid_seed_genes, max_iterations=max_iterations)\n",
    "\n",
    "    # Exclude seed genes from the results\n",
    "    putative_genes = [gene for gene in predicted_genes if gene not in valid_seed_genes]\n",
    "\n",
    "    # Get the top 100 putative disease genes\n",
    "    top_100_putative_genes = putative_genes[:100]\n",
    "\n",
    "    # Save the results\n",
    "    output_file = \"top_100_putative_disease_genes_diamond.csv\"\n",
    "    pd.DataFrame(top_100_putative_genes, columns=[\"Gene\"]).to_csv(output_file, index=False)\n",
    "\n",
    "    print(f\"Top 100 putative disease genes saved to '{output_file}'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the provided script in github!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "E:\\University\\BINM\\DiaMond.py:368: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)\n",
      "  p = float(DIAMOnD_node_info[3])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DIAMOnD(): ignoring 1 of 85 seed genes that are not in the network\n",
      "\n",
      " results have been saved to 'Diamond_result.tsv' \n",
      "\n"
     ]
    }
   ],
   "source": [
    "!python DiaMond.py interactome_lcc_edgelist.tsv 84_disease_genes.tsv 100 1 Diamond_result.tsv\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3.2 EnrichR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E:\\University\\BINM\\EnrichR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Syed Habibul Bashar\\AppData\\Roaming\\Python\\Python39\\site-packages\\IPython\\core\\magics\\osm.py:417: UserWarning: using dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    }
   ],
   "source": [
    "cd E:\\University\\BINM\\EnrichR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing category: GO-BP\n",
      "Processing category: GO-MF\n",
      "Processing category: GO-CC\n",
      "Processing category: Reactome\n",
      "Processing category: KEGG\n",
      "Combined analysis results saved to combined_analysis_results.csv\n"
     ]
    }
   ],
   "source": [
    "def load_and_filter_enrichment(file, p_value_threshold=0.05):\n",
    "    \"\"\"\n",
    "    Load and filter EnrichR results by adjusted p-value.\n",
    "\n",
    "    Args:\n",
    "        file (str): Path to the EnrichR results file.\n",
    "        p_value_threshold (float): Threshold for filtering adjusted p-values.\n",
    "\n",
    "    Returns:\n",
    "        DataFrame: Filtered DataFrame with significant terms.\n",
    "    \"\"\"\n",
    "    # Load the EnrichR results (tab-delimited)\n",
    "    df = pd.read_csv(file, delimiter=\"\\t\")\n",
    "\n",
    "    # Check if required columns exist\n",
    "    if \"Adjusted P-value\" not in df.columns or \"Term\" not in df.columns:\n",
    "        raise ValueError(f\"File {file} must contain 'Adjusted P-value' and 'Term' columns.\")\n",
    "\n",
    "    # Filter by adjusted p-value\n",
    "    filtered_df = df[df[\"Adjusted P-value\"] < p_value_threshold]\n",
    "\n",
    "    return filtered_df[[\"Term\", \"Adjusted P-value\"]]\n",
    "\n",
    "def find_overlap(filtered_df1, filtered_df2):\n",
    "    \"\"\"\n",
    "    Find overlap between two filtered enrichment result DataFrames.\n",
    "\n",
    "    Args:\n",
    "        filtered_df1 (DataFrame): Filtered results for the first dataset.\n",
    "        filtered_df2 (DataFrame): Filtered results for the second dataset.\n",
    "\n",
    "    Returns:\n",
    "        DataFrame: DataFrame containing the overlapping terms.\n",
    "    \"\"\"\n",
    "    # Merge the two DataFrames on the \"Term\" column\n",
    "    overlap = pd.merge(\n",
    "        filtered_df1, filtered_df2, on=\"Term\", suffixes=(\"_original\", \"_putative\")\n",
    "    )\n",
    "    return overlap\n",
    "\n",
    "def process_and_evaluate_all(original_files, putative_files, categories, output_file):\n",
    "    \"\"\"\n",
    "    Process all enrichment files, perform overlap evaluation, and save combined results to a single CSV.\n",
    "\n",
    "    Args:\n",
    "        original_files (list): List of file paths for original disease genes.\n",
    "        putative_files (list): List of file paths for putative disease genes.\n",
    "        categories (list): List of categories (e.g., GO-BP, GO-MF).\n",
    "        output_file (str): Path to save the combined analysis results as a CSV file.\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    # List to store all results\n",
    "    combined_results = []\n",
    "\n",
    "    for original_file, putative_file, category in zip(original_files, putative_files, categories):\n",
    "        print(f\"Processing category: {category}\")\n",
    "\n",
    "        # Load and filter the enrichment results\n",
    "        filtered_original = load_and_filter_enrichment(original_file)\n",
    "        filtered_putative = load_and_filter_enrichment(putative_file)\n",
    "\n",
    "        # Find overlapping terms\n",
    "        overlap = find_overlap(filtered_original, filtered_putative)\n",
    "\n",
    "        # Extract sets of terms\n",
    "        original_terms = set(filtered_original[\"Term\"])\n",
    "        putative_terms = set(filtered_putative[\"Term\"])\n",
    "        overlap_terms = set(overlap[\"Term\"])\n",
    "\n",
    "        # Calculate metrics\n",
    "        num_original_terms = len(original_terms)\n",
    "        num_putative_terms = len(putative_terms)\n",
    "        num_overlap_terms = len(overlap_terms)\n",
    "        percentage_overlap_original = (num_overlap_terms / num_original_terms) * 100 if num_original_terms > 0 else 0\n",
    "        percentage_overlap_putative = (num_overlap_terms / num_putative_terms) * 100 if num_putative_terms > 0 else 0\n",
    "        unique_to_original = original_terms - overlap_terms\n",
    "        unique_to_putative = putative_terms - overlap_terms\n",
    "\n",
    "        # Append results to the combined list\n",
    "        combined_results.append({\n",
    "            \"Category\": category,\n",
    "            \"Total Original Terms\": num_original_terms,\n",
    "            \"Total Putative Terms\": num_putative_terms,\n",
    "            \"Number of Overlapping Terms\": num_overlap_terms,\n",
    "            \"Percentage Overlap (Original)\": percentage_overlap_original,\n",
    "            \"Percentage Overlap (Putative)\": percentage_overlap_putative,\n",
    "            \"Unique Terms in Original\": len(unique_to_original),\n",
    "            \"Unique Terms in Putative\": len(unique_to_putative),\n",
    "        })\n",
    "\n",
    "    # Convert combined results to a DataFrame and save to CSV\n",
    "    combined_results_df = pd.DataFrame(combined_results)\n",
    "    combined_results_df.to_csv(output_file, index=False)\n",
    "    print(f\"Combined analysis results saved to {output_file}\")\n",
    "\n",
    "# Example Usage\n",
    "if __name__ == \"__main__\":\n",
    "    # Define file paths for original and putative results\n",
    "    original_files = [\n",
    "        \"GO_BP_original.txt\",\n",
    "        \"GO_MF_original.txt\",\n",
    "        \"GO_CC_original.txt\",\n",
    "        \"Reactome_original.txt\",\n",
    "        \"KEGG_original.txt\",\n",
    "    ]\n",
    "    putative_files = [\n",
    "        \"GO_BP_putative.txt\",\n",
    "        \"GO_MF_putative.txt\",\n",
    "        \"GO_CC_putative.txt\",\n",
    "        \"Reactome_putative.txt\",\n",
    "        \"KEGG_putative.txt\",\n",
    "    ]\n",
    "\n",
    "    # Define categories\n",
    "    categories = [\"GO-BP\", \"GO-MF\", \"GO-CC\", \"Reactome\", \"KEGG\"]\n",
    "\n",
    "    # Output file for combined results\n",
    "    output_file = \"combined_analysis_results.csv\"\n",
    "\n",
    "    # Process all files and evaluate overlaps\n",
    "    process_and_evaluate_all(original_files, putative_files, categories, output_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Drugs repurposing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E:\\University\\BINM\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Syed Habibul Bashar\\AppData\\Roaming\\Python\\Python39\\site-packages\\IPython\\core\\magics\\osm.py:417: UserWarning: using dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    }
   ],
   "source": [
    "cd E:\\University\\BINM\\"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of gene-drug associations found: 35\n",
      "Results saved to 'Gene_to_Approved_Drugs_With_Status.csv'\n",
      "      gene_name   drug_name   approved\n",
      "773      NME2      LAMIVUDINE   True  \n",
      "2973     NME2       TENOFOVIR   True  \n",
      "11449    CDK2      LOVASTATIN   True  \n",
      "13884     FN1     OCRIPLASMIN   True  \n",
      "14078    CFL1    CLOTRIMAZOLE   True  \n"
     ]
    }
   ],
   "source": [
    "# Load the TSV files\n",
    "interactions_file = \"Interactions.tsv\"\n",
    "first_20_genes_file = \"First_20_Gene.txt\"\n",
    "\n",
    "interactions = pd.read_csv(interactions_file, delimiter=\"\\t\")\n",
    "first_20_genes = pd.read_csv(first_20_genes_file, delimiter=\"\\t\")\n",
    "\n",
    "# Ensure columns exist\n",
    "if \"gene_name\" not in interactions.columns or \"drug_name\" not in interactions.columns or \"approved\" not in interactions.columns:\n",
    "    raise ValueError(\"Interactions file must contain 'gene_name', 'drug_name', and 'approved' columns.\")\n",
    "if \"Gene\" not in first_20_genes.columns:\n",
    "    raise ValueError(\"First_20_Gene file must contain a 'Gene' column.\")\n",
    "\n",
    "# Find genes in \"First_20_Gene\" that are in \"Interactions\"\n",
    "first_20_gene_list = set(first_20_genes[\"Gene\"])\n",
    "filtered_interactions = interactions[interactions[\"gene_name\"].isin(first_20_gene_list)]\n",
    "\n",
    "# Filter for approved drugs (approved=True)\n",
    "approved_drugs = filtered_interactions[filtered_interactions[\"approved\"] == True]\n",
    "\n",
    "# Select relevant columns (gene_name, drug_name, approved)\n",
    "result = approved_drugs[[\"gene_name\", \"drug_name\", \"approved\"]].drop_duplicates()\n",
    "\n",
    "# Save the results to a new CSV file\n",
    "output_file = \"Gene_to_Approved_Drugs_With_Status.csv\"\n",
    "result.to_csv(output_file, index=False)\n",
    "\n",
    "# Print the results\n",
    "print(f\"Number of gene-drug associations found: {len(result)}\")\n",
    "print(f\"Results saved to '{output_file}'\")\n",
    "print(result.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene-drug summary saved to 'Gene_Drug_Summary.csv'\n",
      "  gene_name                  associated_drugs                   drug_count\n",
      "0    BAP1    [PANOBINOSTAT, VORINOSTAT, SUNITINIB, EVEROLIM...       6    \n",
      "1    CDK2    [LOVASTATIN, PACLITAXEL, RESVERATROL, DAUNORUB...       9    \n",
      "2    CFL1    [CLOTRIMAZOLE, CINNARIZINE, CLOFIBRATE, FENOFI...       5    \n",
      "3    EZH2    [TAZEMETOSTAT, DABRAFENIB, TAZEMETOSTAT HYDROB...       3    \n",
      "4     FN1                           [OCRIPLASMIN, DACARBAZINE]       2    \n",
      "5    NME2    [LAMIVUDINE, TENOFOVIR, ZIDOVUDINE, PROGESTERO...       5    \n",
      "6    SOD1    [TETRACYCLINE, DOXYCYCLINE ANHYDROUS, TOFERSEN...       4    \n",
      "7     VCP                                    [HEXACHLOROPHENE]       1    \n"
     ]
    }
   ],
   "source": [
    "# Remove duplicate (gene_name, drug_name) pairs to ensure unique associations\n",
    "approved_drugs_unique = approved_drugs.drop_duplicates(subset=[\"gene_name\", \"drug_name\"])\n",
    "\n",
    "# Group by gene_name and aggregate associated drugs\n",
    "gene_drug_summary = approved_drugs_unique.groupby(\"gene_name\")[\"drug_name\"].apply(list).reset_index(name=\"associated_drugs\")\n",
    "\n",
    "# Add a count column for the number of drugs associated with each gene\n",
    "gene_drug_summary[\"drug_count\"] = gene_drug_summary[\"associated_drugs\"].apply(len)\n",
    "\n",
    "# Save the summary to a CSV file\n",
    "summary_file = \"Gene_Drug_Summary.csv\"\n",
    "gene_drug_summary.to_csv(summary_file, index=False)\n",
    "\n",
    "# Print the summary\n",
    "print(f\"Gene-drug summary saved to '{summary_file}'\")\n",
    "print(gene_drug_summary)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. ProConSuL vs DIAMOnD (Best performer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PROCONSUL(): ignoring 1 of 48 seed genes that are not in the network\n",
      "PROCONSUL(): Round 1/10\n",
      "PROCONSUL(): Round 2/10\n",
      "PROCONSUL(): Round 3/10\n",
      "PROCONSUL(): Round 4/10\n",
      "PROCONSUL(): Round 5/10\n",
      "PROCONSUL(): Round 6/10\n",
      "PROCONSUL(): Round 7/10\n",
      "PROCONSUL(): Round 8/10\n",
      "PROCONSUL(): Round 9/10\n",
      "PROCONSUL(): Round 10/10\n",
      "\n",
      " results have been saved to 'ProConSuL_results.csv' \n",
      "\n"
     ]
    }
   ],
   "source": [
    "!python proconsul.py --network_file interactome_lcc_edgelist.tsv --seed_file 84_disease_genes.tsv --n 20 --outfile_name ProConSuL_results.csv --temp 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "processing the outputfile for comarision with the DIAMOnD nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Node column saved to ProConSuL_results_Nodes.csv\n"
     ]
    }
   ],
   "source": [
    "# Load the data (adjust the file name and delimiter as needed)\n",
    "file_path = \"ProConSuL_results.csv\"  # Replace with your file name\n",
    "data = pd.read_csv(file_path, delimiter=\"\\t\")\n",
    "\n",
    "# Extract the 'node' column\n",
    "node_column = data[\"node\"]\n",
    "\n",
    "# Save the node column to a new file\n",
    "output_file = \"ProConSuL_results_Nodes.csv\"\n",
    "node_column.to_csv(output_file, index=False)\n",
    "\n",
    "# Print success message\n",
    "print(f\"Node column saved to {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overlap: {'ACO2', 'VCP', 'SOD1', 'CFL1', 'CDK2', 'EZH2', 'CLIC1', 'CLIC4', 'BAP1', 'DSTN', 'UBE2M', 'PRDX6', 'AGR2', 'PRDX2'}\n",
      "Unique to ProConSuL: {'U2AF2', 'HSPA8', 'MCM2', 'ISG15', 'RICTOR', 'IQGAP1'}\n",
      "Unique to DIAMOnD: {'NME2', 'FN1', 'LDHA', 'TKT', 'MYOC', 'TIMP2'}\n"
     ]
    }
   ],
   "source": [
    "# Load ProConSuL results\n",
    "proconsul_results_file = \"ProConSuL_results_Nodes.csv\"  # Replace with your actual file\n",
    "proconsul_results = pd.read_csv(proconsul_results_file)\n",
    "top_20_proconsul_genes = proconsul_results.head(20)['node'].tolist()\n",
    "\n",
    "# Load DIAMOnD results\n",
    "diamond_results_file = \"100_putative_gene.tsv\"  # Use the DIAMOnD top 20 file\n",
    "diamond_results = pd.read_csv(diamond_results_file)\n",
    "top_20_diamond_genes = diamond_results.head(20)['DIAMOnD_node'].tolist()\n",
    "\n",
    "# Compare the two lists\n",
    "overlap = set(top_20_proconsul_genes).intersection(set(top_20_diamond_genes))\n",
    "unique_to_proconsul = set(top_20_proconsul_genes) - set(top_20_diamond_genes)\n",
    "unique_to_diamond = set(top_20_diamond_genes) - set(top_20_proconsul_genes)\n",
    "\n",
    "# Print results\n",
    "print(\"Overlap:\", overlap)\n",
    "print(\"Unique to ProConSuL:\", unique_to_proconsul)\n",
    "print(\"Unique to DIAMOnD:\", unique_to_diamond)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
